Static to static calls (#464)
powdr-labs/powdr: analysis/README.md

## Definitions

We define two types of machines: virtual machines and constrained machines. Constrained machines are the lower level kind of machine. They have a notion of blocks through a latch and an operation_id. Virtual machines are a higher level type of machines. For each machine type, we provide the elements which only appear exclusively in that type.

|                         | Virtual | Constrained |
|-------------------------|---------|-------------|
| pc                      | yes     | no          |
| latch                   | no      | yes         |
| operation_id            | no      | yes         |
| functions               | yes     | no          |

> | instructions            | yes     | no          |

Author: Good point but no, links are allowed in VMs

> | instructions            | yes     | no          |

Reviewer: ```suggestion
| instructions            | yes     | no          |
| links                   | yes     | yes         |
```

This way then? ;) I think links should be mentioned here, otherwise it's the only major concept that's not.

- Find the maximum number of inputs among all functions. Introduce as many input registers. Do the same for outputs, introducing output registers
- Replace references to the function arguments by references to these input registers.
- Pad all return statements with zeroes up to the number of output registers.

> - Inline all functions arbitrarily ordered by name, adding a label before each one.

Reviewer 2: ```suggestion
- Inline all function bodies, arbitrarily sorted by name, adding a label before each one.
```
Otherwise it's not clear to me what "inline all functions" mean

The final program after analysis is the following:
```
machine DifferentSignatures(instr_return, _operation_id) {
        operation identity<2> _input_0 -> _output_0;
        operation nothing<3>;

>         operation one<4>  -> _output_0;

Reviewer 2: ```suggestion
        operation one<4> -> _output_0;
```
was this auto generated?

Author: yes

        constraints {
                pol constant _block_enforcer_last_step = [0]* + [1];
                pol commit _operation_id_no_change;
                _operation_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
                (_operation_id_no_change * (_operation_id' - _operation_id)) = 0;
        }

        link instr_identity  X -> Y = sub.identity;
        link instr_one   -> Y = sub.one;
        link instr_nothing  = sub.nothing;

}
```

> 

Author: fixed

pol constant p_reg_write_Y_A = [0, 0, 1, 0, 0] + [0]*;
{ pc, reg_write_X_A, reg_write_Y_A, instr_identity, instr_one, instr_nothing, instr__jump_to_operation, instr__reset, instr__loop, instr_return, X_const, X_read_free, read_X_A, read_X_pc, Y_const, Y_read_free, read_Y_A, read_Y_pc } in { p_line, p_reg_write_X_A, p_reg_write_Y_A, p_instr_identity, p_instr_one, p_instr_nothing, p_instr__jump_to_operation, p_instr__reset, p_instr__loop, p_instr_return, p_X_const, p_X_read_free, p_read_X_A, p_read_X_pc, p_Y_const, p_Y_read_free, p_read_Y_A, p_read_Y_pc };
pol constant _block_enforcer_last_step = [0]* + [1];
pol commit _operation_id_no_change;
_operation_id_no_change = ((1 - _block_enforcer_last_step) * (1 - instr_return));
(_operation_id_no_change * (_operation_id' - _operation_id)) = 0;
// Links:
// instr_identity  X -> Y links to function "identity" with id 2 with params  _input_0 -> _output_0 in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"
// instr_one   -> Y links to function "one" with id 4 with params   -> _output_0 in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"

> // instr_nothing  links to function "nothing" with id 3 with params  in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"

Author: There is no syntax for links in object files, so they are just referred to as comments. They are not expressible as PIL code unlike the rest of the object.

> // instr_nothing  links to function "nothing" with id 3 with params  in object at location "main_sub" with latch "instr_return" and operation_id "_operation_id"

Author: Lookups are one way to encode the links. Links are turned into lookups in the linker. Here we are before the linker, where we have a bunch of independent tables, and some links between them, but no specific implementation of the links.

```
namespace main(16);

> pol commit _operation_id;

Author: #514
